gantry/routes/prediction/prediction.py

```diff
@@ -0,0 +1,146 @@
+import logging
+
+import aiosqlite
+
+from gantry.routes.prediction.current_mapping import pkg_mappings
+from gantry.util.k8s import convert_bytes, convert_cores
+
+IDEAL_SAMPLE = 5
+DEFAULT_CPU_REQUEST = 1.0
+DEFAULT_MEM_REQUEST = 2000 * 1_000_000  # 2GB in bytes
+
+
+async def predict_single(db: aiosqlite.Connection, build: dict) -> dict:
+    """
+    Predict the resource usage of a build
+
+    args:
+        build: dict that must contain pkg name, pkg version, compiler, compiler version
+    returns:
+        dict of predicted resource usage: cpu_request, mem_request
+        CPU in millicore, mem in MB
+    """
+
+    sample = await get_sample(db, build)
+    predictions = {}
+    if not sample:
+        predictions = {
+            "cpu_request": DEFAULT_CPU_REQUEST,
+            "mem_request": DEFAULT_MEM_REQUEST,
+        }
+    else:
+        # mapping of sample: [0] cpu_mean, [1] cpu_max, [2] mem_mean, [3] mem_max
+        predictions = {
+            # averages the respective metric in the sample
+            "cpu_request": sum([build[0] for build in sample]) / len(sample),
+            "mem_request": (sum([build[2] for build in sample]) / len(sample)),
+        }
+
+    ensure_higher_pred(predictions, build["package"]["name"])
+
+    # warn if the prediction is below some thresholds
+    if predictions["cpu_request"] < 0.25:
+        logging.warning(f"Warning: CPU request for {build['hash']} is below 0.25 cores")
+        predictions["cpu_request"] = DEFAULT_CPU_REQUEST
+    if predictions["mem_request"] < 10_000_000:
+        logging.warning(f"Warning: Memory request for {build['hash']} is below 10MB")
+        predictions["mem_request"] = DEFAULT_MEM_REQUEST
+
+    # convert predictions to k8s friendly format
+    for k, v in predictions.items():
+        if k.startswith("cpu"):
+            predictions[k] = convert_cores(v)
+        elif k.startswith("mem"):
+            predictions[k] = convert_bytes(v)
+
+    return {
+        "hash": build["hash"],
+        "variables": {
+            # spack uses these env vars to set the resource requests
+            # set them here at the last minute to avoid using these vars
+            # and clogging up the code
+            "KUBERNETES_CPU_REQUEST": predictions["cpu_request"],
+            "KUBERNETES_MEMORY_REQUEST": predictions["mem_request"],
+        },
+    }
+
+
+async def predict_bulk(db: aiosqlite.Connection, builds: list) -> list:
+    """
+    Handles a bulk request of builds
+
+    args:
+        builds: list of dicts (see predict_single)
+    returns: see predict_single
+    """
+
+    return [await predict_single(db, build) for build in builds]
```

Comment:
Do you happen to know if this executes in parallel? If not, we could look into asyncio task pools if this gets slow.

Reply:
Great catch! this should work now with [a2aa3db](https://github.com/spack/spack-gantry/pull/5/commits/a2aa3dbfbbf7b7aa5fc0234cf3db0fddcda91916)